[docs] Add SSR section (#880)
lit/lit.dev: packages/lit-dev-content/site/docs/ssr/authoring.md

```

The Lit team is also working on providing an easy way of checking whether the code is running in a Node environment which can be utilized to write conditional code blocks targeting different environments. Follow the issue [[labs/ssr] A way for client code to check if running in server (isSsr)](https://github.com/lit/lit/issues/3158) for updates on this feature.

For more complex uses cases, consider utilizing [conditional exports](https://nodejs.org/api/packages.html#conditional-exports) in Node that specifically match for `"node"` environments so you could have different code depending on whether the module is being imported for use in Node or in the browser. Users would get the appropriate version of the package depending on whether it was imported from Node or the browser. Export conditions are also supported by popular bundling tools like [rollup](https://github.com/rollup/plugins/tree/master/packages/node-resolve#exportconditions) and [webpack](https://webpack.js.org/configuration/resolve/#resolveconditionnames) so users can bring in the appropriate code for your bundle.

{% aside "warn" %}

Don't bundle Lit into published components.

> Due to the way Lit libraries utilize conditional export for providing code meant for import in Node, we strongly discourage bundling `lit` into your component for packages being published to NPM as your bundle will only include a particular version of `lit` meant for a particular environment.

Reviewer: ```suggestion
Because Lit packages use [conditional exports](https://nodejs.org/api/packages.html#conditional-exports) to provide different modules to Node and browser environments, we strongly discourage bundling `lit` into your packages being published to NPM. If you do, your bundle will only include `lit` modules meant for the environment you bundled, and won't automatically switch based on environment.
```

Author: The link to nodejs conditional exports doc is already present in the paragraph immediately above this warning so I don't think we need to link it here again, but the rephrasing is great!

---
title: Authoring components for Lit SSR
eleventyNavigation:
  key: Authoring components
  parent: Server rendering
  order: 4
---

{% labs-disclaimer %}

> Lit's approach to rendering web components in a server environment places some restrictions on component code to achieve efficient server rendering. When authoring components, keep in mind these considerations if you your components to be compatible with Lit SSR.
> 

Reviewer 2: ```suggestion
Lit's approach to rendering web components in a server environment places some restrictions on component code to achieve efficient server rendering. When authoring components, keep in mind these considerations to ensure they are compatible with Lit SSR.

```

Note: The restrictions listed on this page are subject to change as we make improvements to Lit SSR. If you would like to see a certain use case supported, please [file an issue](https://github.com/lit/lit/issues/new/choose) or start a [discussion](https://github.com/lit/lit/discussions) thread.

## Considerations

### Browser only code

Most browser DOM APIs are not available in the Node environment. Lit SSR utilizes a DOM shim that's the bare minimum required for rendering Lit templates and components. For a full list of what APIs are available, see the [DOM Emulation](/docs/ssr/dom-emulation) page.

When authoring components, perform imperative DOM operations from lifecycle methods that are called only on the client, and not on the server. See the [lifecycles](#lifecycles) section below for what specific methods can be used.

> Some component modules may also have side effects that utilize browser APIs, to detect for certain browser features for example, which would break simply by importing in a non-browser environment. These side effects, if they cannot be moved to a client-only lifecycle, must have proper alternatives or some default behavior for when running in Node.

Reviewer 2: ```suggestion
Some component modules may also have side effects that utilize browser APIs, for example to detect certain browser features, which would break simply by importing in a non-browser environment. If these side-effects cannot be moved to a client-only lifecycle, you must provide proper alternatives or some default behavior for when running in Node.
```

> For simple cases, adding conditionals or optional chaining to certain DOM accesses may be sufficient to guard against unavailable DOM APIs, such as

Reviewer 2: ```suggestion
For simple cases, adding conditionals or optional chaining to certain DOM accesses may be sufficient to guard against unavailable DOM APIs. For example:
```

---
title: Authoring components for Lit SSR
eleventyNavigation:
  key: Authoring components
  parent: Server rendering
  order: 4
---

{% labs-disclaimer %}

Lit's approach to rendering web components in a server environment places some restrictions on component code to achieve efficient server rendering. When authoring components, keep in mind these considerations to ensure they are compatible with Lit SSR.

Note: The restrictions listed on this page are subject to change as we make improvements to Lit SSR. If you would like to see a certain use case supported, please [file an issue](https://github.com/lit/lit/issues/new/choose) or start a [discussion](https://github.com/lit/lit/discussions) thread.

> ## Considerations

Reviewer 3: As a point of style, we try to avoid empty headings (that is, a heading followed immediately by another heading with no text in between.

In this case, I'm not sure this grouping adds anything, so I'd recommend cutting it.

> ### Browser only code

Reviewer 3: ```suggestion
### Browser-only code
```

Most browser DOM APIs are not available in the Node environment. Lit SSR utilizes a DOM shim that's the bare minimum required for rendering Lit templates and components. For a full list of what APIs are available, see the [DOM Emulation](/docs/ssr/dom-emulation) page.

> When authoring components, perform imperative DOM operations from lifecycle methods that are called only on the client, and not on the server. See the [lifecycle](#lifecycle) section below for what specific methods can be used.

Reviewer 3: It would be really nice here to include even a high-level example. 

```suggestion
When authoring components, perform imperative DOM operations from lifecycle methods that are called **only on the client**, and not on the server. For example, use `updated()` if you need to measure updated DOM. This callback is only run on the browser, so it's safe to access the DOM.

See the [lifecycle](#lifecycle) section below for a list of which methods are called on the server and which are browser-only.
```

Reviewer 3: You might have a better example, here, just throwing something out.

Author: Thank you that's a great example!

> Some component modules may also have side effects that utilize browser APIs, for example to detect certain browser features, which would break simply by importing in a non-browser environment. If these side-effects cannot be moved to a client-only lifecycle, you must provide proper alternatives or some default behavior for when running in Node.

Reviewer 3: This paragraph confused me. There's a lot packed in here, plus several pretty vague references (like "proper alternatives").

I think you're saying, roughly: "You might have a module that defines a Lit component, but _also_ has a side effect that could break on import. In this case, you have a few choices:

- Move the side effect code into a Lit lifecycle callback that's only called on the browser.
- Conditionalize the side-effect code such that it doesn't run on the server.
- I'm actually not sure what else "property alternatives or some default behavior" might involve."

If so, here's my first stab at a rephrase.

```suggestion
Some modules that define Lit components may also have side effects that use browser APIs—for example to detect certain browser features—such that the module breaks when imported in a non-browser environment. In this case, you can move the side effect code into a browser-only lifecycle callback, or conditionalize so that it only runs on the browser.
```

Also given that the section is called "Browser-only code," consider standardizing on "browser-only" (like, "browser-only callback" maybe) instead of switching between "browser" and "client" which can make things a little less clear.

```

The Lit team is also working on providing an easy way of checking whether the code is running in a Node environment which can be utilized to write conditional code blocks targeting different environments. Follow the issue [[labs/ssr] A way for client code to check if running in server (isSsr)](https://github.com/lit/lit/issues/3158) for updates on this feature.

For more complex uses cases, consider utilizing [conditional exports](https://nodejs.org/api/packages.html#conditional-exports) in Node that specifically match for `"node"` environments so you could have different code depending on whether the module is being imported for use in Node or in the browser. Users would get the appropriate version of the package depending on whether it was imported from Node or the browser. Export conditions are also supported by popular bundling tools like [rollup](https://github.com/rollup/plugins/tree/master/packages/node-resolve#exportconditions) and [webpack](https://webpack.js.org/configuration/resolve/#resolveconditionnames) so users can bring in the appropriate code for your bundle.

> Note: Due to the way Lit libraries utilize conditional export for providing code meant for import in Node, we strongly discourage bundling `lit` into your component for packages being published to NPM as your bundle will only include a particular version of `lit` meant for a particular environment.

Reviewer 3: I would put this in an aside and up it to a warning.

```suggestion
{% aside "warning" %}

Don't bundle Lit into published components.

Due to the way Lit libraries utilize conditional export for providing code meant for import in Node, we strongly discourage bundling `lit` into your component for packages being published to NPM as your bundle will only include a particular version of `lit` meant for a particular environment.

{% endaside %}
```

### Lifecycle

> 

Reviewer: I wonder if we should add warning here that this set of server lifecycle callbacks is subject to change while SSR is in labs.

Author: I have that note specifically for `connectedCallback()` only, but that makes it seem like that's the only thing that could change so a generic warning might be better.

Author: I added the warning to the same aside mentioned https://github.com/lit/lit.dev/pull/880#discussion_r954342904 about methods called on server side should not contain DOM APIs. But maybe it's bad to say methods not called on server are safe, immediately followed by this is subject to change?

Reviewer: I think these should be two separate messages. One is permanent, and not a warning: don't call DOM APIs in server lifecycle methods. The other is temporary and a warning: the set of client vs server lifecycle methods might chance while this library is unstable.

Author: Split it up and put the "subject to change" in a labs aside as that's meant to be the temporary one.

> Only a select number of web component, `LitElement`, and Lit directive lifecycle callbacks will be run during server-side rendering that are utilized to generate the initial styling and markup for the component. Additional lifecycle methods will be called client-side during hydration and subsequent changes to the page.

Reviewer 3: Going to recommend edits here for brevity and clarity.

```suggestion
Only certain lifecycle callbacks are run during server-side rendering. These callback generate the initial styling and markup for the component. Additional lifecycle methods are called client-side during hydration and during runtime after the components are hydrated.
```

"changes to the page" sounds like it could mean changes sent from the server, I think we're specifically talking about things that happen on the client side post hydration. Please rephrase if this doesn't seem good.

> The tables below lists the standard custom element and Lit element lifecycle methods and whether they are called during SSR and during hydration.

Reviewer 3: This sentence says, "and during hydration," but the tables don't mention hydration at all.

```suggestion
The tables below lists the standard custom element and Lit lifecycle methods and whether they are called during SSR. All of the methods can be called in the browser.
```

One additional, higher-effort possibility would be to *reverse* the sense of the tables. Instead of "Called on server". you could have "Browser-only" with "Yes" and "No*" values. it would require a little rewriting elsewhere (including the suggestion above) but it would mesh nicely with the section above talking about browser-only callbacks. Don't know if it would help any with the concerns @kevinpschaaf expressed.

Reviewer 4: Just made a suggestion here: https://github.com/lit/lit.dev/pull/880#discussion_r954342904

Reviewer 4: Also might make this a little more precise:
```suggestion
The tables below lists the standard custom element and Lit lifecycle methods and whether they are called during SSR. All of the lifecycle is available on the browser after element registration and hydration.
```
